query-algebrizer/src/cc.rs

```diff
@@ -209,14 +213,31 @@ pub struct ConjoiningClauses {
     pub wheres: Vec<ColumnConstraint>,
 
     /// A map from var to qualified columns. Used to project.
-    pub bindings: BTreeMap<Variable, Vec<QualifiedAlias>>,
+    pub column_bindings: BTreeMap<Variable, Vec<QualifiedAlias>>,
+
+    /// A list of variables mentioned in the query's :in clause. These must all be bound before the
+    /// query can be executed.
+    pub input_variables: BTreeSet<Variable>,
+
+    /// In some situations -- e.g., when a query is being run only once -- we know the values of
+    /// some or all bindings in advance. These can be substituted directly when the query is
+    /// algebrized.
+    ///
+    /// Known bindings must agree with `known_types`. If you write a query like
+    /// ```edn
+    /// [:find ?x :in $ ?val :where [?x :foo/int ?val]]
+    /// ```
+    ///
+    /// and for `?val` provide `TypedValue::String("foo".to_string())`, the query will be known at
+    /// algebrizing time to be empty.
+    known_bindings: BTreeMap<Variable, TypedValue>,
```

Comment:
I think I agree with your naming suggestion. I might also opt for `value_bindings` (cf `column_bindings`).

From the perspective of these being used to bind inputs, yes, `input_variables` should be a superset. The cljs impl doesn't have an analogue of `input_variables`.

The only doubt I have is what to do with nested CCs (e.g., `not` clauses). For an 'inner' CC, it's not obvious what its inputs are.

In the cljs implementation we merge, as we descend, the outer CC's external bindings (mappings from var to expressions like `(sql/param :foo)`) with its column bindings to form the inner CC's external bindings. That allows us to cleanly merge on the way back up: we don't contribute external bindings to the parent.

I haven't got to that point in the Rust codebase yet.